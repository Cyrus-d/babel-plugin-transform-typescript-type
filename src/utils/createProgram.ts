import * as ts from 'typescript';
import { getCompilerOptions, sourceFileCacheInstance } from './SourceFileCache';

export const createProgram = (filePath: string) => {
  const tsconfig = getCompilerOptions();
  const program = ts.createProgram([filePath], tsconfig, {
    fileExists(fileName): boolean {
      const file = sourceFileCacheInstance.createOrUpdateSourceFile(fileName);
      if (!file) {
        return false;
      }

      return true;
    },
    getCanonicalFileName(_fileName: string): string {
      return _fileName;
    },
    getCurrentDirectory(): string {
      return '';
    },
    getDefaultLibFileName(): string {
      return ts.getDefaultLibFileName(tsconfig);
    },

    getNewLine(): string {
      return '\r\n';
    },
    getSourceFile(fileName): ts.SourceFile | undefined {
      return sourceFileCacheInstance.createOrUpdateSourceFile(fileName);
    },
    readFile(fileName: string): string {
      const src = sourceFileCacheInstance.createOrUpdateSourceFile(fileName);
      if (!src) return '';

      return src.text;
    },
    useCaseSensitiveFileNames(): boolean {
      return false;
    },
    writeFile: (_fileName, _content) => null,
  });

  return program;
};
