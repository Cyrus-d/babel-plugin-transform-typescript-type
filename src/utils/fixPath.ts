export const fixPath = (fileName: string) => fileName.split('/').join('\\');
