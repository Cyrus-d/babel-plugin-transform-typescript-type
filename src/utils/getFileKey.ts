export const getFileKey = (fileName: string) =>
  fileName
    .split('/')
    .join('_')
    .split('\\')
    .join('_');
