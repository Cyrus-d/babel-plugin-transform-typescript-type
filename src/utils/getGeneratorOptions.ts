import { types as t } from '@babel/core';
import { GeneratorOptions } from '../typings';

interface Option {
  [key: string]: any;
}

const setArrayOption = (prop: t.ObjectProperty, opts: Option) => {
  if (!t.isArrayExpression(prop.value)) return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.elements.reduce((arr, elm) => {
      if (t.isStringLiteral(elm)) arr.push(elm.value);

      return arr;
    }, [] as string[]);
  }
};

const setOtherOptions = (prop: t.ObjectProperty, opts: Option) => {
  if (
    !t.isNumericLiteral(prop.value) &&
    !t.isStringLiteral(prop.value) &&
    !t.isBooleanLiteral(prop.value)
  )
    return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.value;
  }
};

export const getGeneratorOptions = <T extends t.CallExpression>(node: T): GeneratorOptions<T> => {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  if (!node.arguments) return {} as GeneratorOptions<T>;

  return node.arguments.reduce((opts, arg) => {
    if (t.isObjectExpression(arg)) {
      arg.properties.forEach(prop => {
        if (t.isObjectProperty(prop)) {
          setArrayOption(prop, opts);
          setOtherOptions(prop, opts);
        }
      });
    }

    return opts;
  }, {} as GeneratorOptions<{}>);
};
