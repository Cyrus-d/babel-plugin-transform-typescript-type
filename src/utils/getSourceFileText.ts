import * as fs from 'fs';

export const getSourceFileText = (fileName: string) => {
  if (!fs.existsSync(fileName)) return null;

  return fs.readFileSync(fileName, 'utf8');
};
