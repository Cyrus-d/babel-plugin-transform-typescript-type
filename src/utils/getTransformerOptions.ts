import { types as t } from '@babel/core';
import { TransformerOptions } from '../typings';

interface Option {
  [key: string]: any;
}

const setArrayOption = (prop: t.ObjectProperty, opts: Option) => {
  if (!t.isArrayExpression(prop.value)) return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.elements.reduce((arr, elm) => {
      if (t.isStringLiteral(elm)) arr.push(elm.value);

      return arr;
    }, [] as string[]);
  }
};

const setOtherOptions = (prop: t.ObjectProperty, opts: Option) => {
  if (
    !t.isNumericLiteral(prop.value) &&
    !t.isStringLiteral(prop.value) &&
    !t.isBooleanLiteral(prop.value)
  )
    return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.value;
  }
};

export const getTransformerOptions = <T extends t.CallExpression>(
  node: T,
): TransformerOptions<T> => {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  if (!node.arguments) return {} as TransformerOptions<T>;

  return node.arguments.reduce((opts, arg) => {
    if (t.isObjectExpression(arg)) {
      arg.properties.forEach(prop => {
        if (t.isObjectProperty(prop)) {
          setArrayOption(prop, opts);
          setOtherOptions(prop, opts);
        }
      });
    }

    return opts;
  }, {} as TransformerOptions<{}>);
};
