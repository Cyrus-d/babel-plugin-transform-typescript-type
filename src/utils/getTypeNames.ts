import { types as t } from '@babel/core';

const getTSTypeReferenceTypeName = (node: t.TSTypeReference) => {
  if (t.isIdentifier(node.typeName)) {
    return node.typeName.name;
  }

  return null;
};

const getIntersectionTypeNames = (node: t.TSIntersectionType) => {
  const typeNames = node.types.reduce((arr, type) => {
    if (t.isTSTypeReference(type)) {
      const typeName = getTSTypeReferenceTypeName(type);
      if (typeName) arr.push(typeName);
    }

    return arr;
  }, [] as string[]);

  return typeNames;
};

export const getTsTypeName = (node: t.TSType) => {
  let typeNames: string[] = [];
  if (t.isTSTypeReference(node)) {
    const typeName = getTSTypeReferenceTypeName(node);
    if (typeName) typeNames.push(typeName);
  } else if (t.isTSIntersectionType(node)) {
    typeNames = getIntersectionTypeNames(node);
  }

  return typeNames;
};

const getTypeParametersTypeNames = (typeParameters: t.TSTypeParameterInstantiation) => {
  return typeParameters.params.reduce((arr, param) => {
    return getTsTypeName(param);
  }, [] as string[]);
};

export const getNodeTypesNames = (node: t.Node) => {
  if (t.isCallExpression(node)) {
    return getTypeParametersTypeNames((node as any).typeParameters);
  }

  return [];
};
