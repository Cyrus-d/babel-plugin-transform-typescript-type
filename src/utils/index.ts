export * from './createProgram';
export * from './getFileKey';
export * from './getTransformerOptions';
export * from './getTypeNames';
export * from './mergeSchema';
export * from './moduleDependencies';
export * from './sourceFile';
export * from './updateReferences';
