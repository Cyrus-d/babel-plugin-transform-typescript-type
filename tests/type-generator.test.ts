import path from 'path';
import glob from 'fast-glob';
import { transform } from './utils';

jest.mock('chokidar', () => ({ watch: jest.fn(() => ({ on: jest.fn(() => {}) })) }));

describe('type-generator', () => {
  glob
    .sync('./fixtures/manual-converter/type-schema/**/*.{ts,tsx}', { cwd: __dirname, dot: false })
    .forEach((basePath) => {
      const filePath = String(basePath);

      if (filePath.includes('/special/')) {
        return;
      }
      it(`transforms ${filePath}`, () => {
        expect(
          transform(path.join(__dirname, filePath), {}, { generateReactPropTypesManually: true }),
        ).toMatchSnapshot();
      });
    });

  it('should not transforms when disabled by function', () => {
    expect(
      transform(
        path.join(
          __dirname,
          './fixtures/manual-converter/type-schema/special/schema-function-disable.ts',
        ),
      ),
    ).toMatchSnapshot();
  });

  it('should not transforms when disabled globally', () => {
    expect(
      transform(
        path.join(
          __dirname,
          './fixtures/manual-converter/type-schema/special/skipParseFiles-defaults.ts',
        ),
        {},
        {
          disableGenerateTypeSchemaInEnv: ['test'],
        },
      ),
    ).toMatchSnapshot();
  });

  it('should merge skipParseTypes items', () => {
    expect(
      transform(
        path.join(
          __dirname,
          './fixtures/manual-converter/type-schema/special/skipParseTypes-defaults.ts',
        ),
        {},
        {
          skipParseTypes: ['HTMLElement'],
        },
      ),
    ).toMatchSnapshot();
  });

  it('should merge skipParseFiles items', () => {
    expect(
      transform(
        path.join(
          __dirname,
          './fixtures/manual-converter/type-schema/special/skipParseFiles-defaults.ts',
        ),
        {},
        {
          skipParseFiles: ['lib.dom.iterable.d.ts'],
        },
      ),
    ).toMatchSnapshot();
  });
});
