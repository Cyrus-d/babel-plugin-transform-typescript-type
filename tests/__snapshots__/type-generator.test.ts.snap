// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type-generator should merge skipParseFiles items 1`] = `
"import { transformTypeToSchema } from '../../../../../src/transformerFunctions';
export interface TestProps {
  a: HTMLElement;
  b?: AudioParam;
}
const type = {
  'type': 'object',
  'properties': {
    'a': {
      'type': 'HTMLElement'
    },
    'b': {
      'type': 'AudioParam'
    }
  },
  'required': ['a'],
  'additionalProperties': false
};"
`;

exports[`type-generator should merge skipParseTypes items 1`] = `
"import { transformTypeToSchema } from '../../../../../src/transformerFunctions';
export interface TestProps {
  a: HTMLElement;
  b?: HTMLDivElement;
}
const type = {
  'type': 'object',
  'properties': {
    'a': {
      'type': 'HTMLElement'
    },
    'b': {
      'type': 'HTMLDivElement'
    }
  },
  'required': ['a'],
  'additionalProperties': false
};"
`;

exports[`type-generator should not transforms when disabled by function 1`] = `
"import { transformTypeToSchema } from '../../../../../src/transformerFunctions';
export interface TestProps {
  prop_a: string;
  prop_b?: string;
}
const shouldBeNullValue = null;"
`;

exports[`type-generator should not transforms when disabled globally 1`] = `
"import { transformTypeToSchema } from '../../../../../src/transformerFunctions';
export interface TestProps {
  a: HTMLElement;
  b?: AudioParam;
}
const type = null;"
`;

exports[`type-generator transforms ./fixtures/manual-converter/type-schema/extended-type-schema.ts 1`] = `
"import { ExternalProps } from '../typings';
import { transformTypeToSchema } from '../../../../src/transformerFunctions';
export interface TestProps extends ExternalProps {
  prop_a: string;
  prop_b?: string;
}
const type = {
  'type': 'object',
  'properties': {
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    },
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'string'
    }
  },
  'required': ['ext_prop_1', 'ext_prop_2', 'prop_a'],
  'additionalProperties': false
};"
`;

exports[`type-generator transforms ./fixtures/manual-converter/type-schema/interface-schema.ts 1`] = `
"import { transformTypeToSchema } from '../../../../src/transformerFunctions';
export interface TestProps {
  prop_a: string;
  prop_b?: string;
}
const type = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'string'
    }
  },
  'required': ['prop_a'],
  'additionalProperties': false
};
const type2 = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    }
  },
  'required': ['prop_a'],
  'additionalProperties': false
};
const type3 = {
  'type': 'object',
  'properties': {
    'prop_b': {
      'type': 'string'
    }
  },
  'additionalProperties': false
};"
`;

exports[`type-generator transforms ./fixtures/manual-converter/type-schema/interface-schema-maxDepth.ts 1`] = `
"import { transformTypeToSchema } from '../../../../src/transformerFunctions';
export interface TestProps {
  level_1: {
    level_2: {
      level_3: string;
    };
  };
}
const type4 = {
  'type': 'object',
  'properties': {
    'level_1': {
      'type': 'object',
      'properties': {
        'level_2': {
          'type': 'object',
          'additionalProperties': false
        }
      },
      'required': ['level_2'],
      'additionalProperties': false
    }
  },
  'required': ['level_1'],
  'additionalProperties': false
};"
`;

exports[`type-generator transforms ./fixtures/manual-converter/type-schema/mix-type-schema.ts 1`] = `
"import { ExternalProps } from '../typings';
import { transformTypeToSchema } from '../../../../src/transformerFunctions';
export interface TestProps {
  prop_a: string;
  prop_b?: string;
}
const type = {
  'type': 'object',
  'properties': {
    'prop_a': {
      'type': 'string'
    },
    'prop_b': {
      'type': 'string'
    },
    'ext_prop_1': {
      'type': 'string'
    },
    'ext_prop_2': {
      'type': 'string'
    }
  },
  'required': ['prop_a', 'ext_prop_1', 'ext_prop_2'],
  'additionalProperties': false
};"
`;

exports[`type-generator transforms ./fixtures/manual-converter/type-schema/skipParseTypeInFiles.ts 1`] = `
"import { ExternalProps } from '../typings';
import { transformTypeToSchema } from '../../../../src/transformerFunctions';
export interface TestProps {
  a: HTMLElement;
  b: string;
}
const type = {
  'type': 'object',
  'properties': {
    'a': {
      'type': 'HTMLElement'
    },
    'b': {
      'type': 'string'
    }
  },
  'required': ['a', 'b'],
  'additionalProperties': false
};"
`;

exports[`type-generator transforms ./fixtures/manual-converter/type-schema/skipParseTypes.ts 1`] = `
"import { ExternalProps } from '../typings';
import { transformTypeToSchema } from '../../../../src/transformerFunctions';
export interface TestProps {
  a: HTMLElement;
  b: string;
}
const type = {
  'type': 'object',
  'properties': {
    'a': {
      'type': 'HTMLElement'
    },
    'b': {
      'type': 'string'
    }
  },
  'required': ['a', 'b'],
  'additionalProperties': false
};"
`;
